### DISCUSSION TITLE
How to execute nextflow some process parallel and some process sequential way please give me suggestion

### DISCUSSION THREAD

**Discussion by satish15625** [nextflow-io/nextflow#3400] [General]
/*
NXF ver 19.08+ needed because of the use of tuple instead of set
*/
nextflow.enable.dsl=2
if( !nextflow.version.matches('>=19.08') ) {
    println "This workflow requires Nextflow version 19.08 or greater and you are running version $nextflow.version"
    exit 1
}

/*
* ANSI escape codes to color output messages, get date to use in results folder name
*/
ANSI_GREEN = "\033[1;32m"
ANSI_RED = "\033[1;31m"
ANSI_RESET = "\033[0m"

/* pipeline input parameters 
*/
params.saveMode = 'copy'
params.readsdir =""
params.outdir = "${workflow.launchDir}/results"
params.fqpattern = "*_{,R}{1,2}{,_001}.{fq,fastq}{,.gz}"
params.ontreads=false
/*
    reference files
*/
params.references = "$baseDir/references"
/* Pipeline Title 
params.title = "SNP Detection Pipeline."
*/
/*
modules path conda env  
*/
params.modules = "$baseDir/lib"

params.help = ""

if (params.help) {
    helpMessage()
    exit(0)
}

log.info """
        ===========================================
         SNP DETECTION P I P E L I N E    

         Used parameters:
        -------------------------------------------
         --readsdir         : ${params.readsdir}
         --fqpattern        : ${params.fqpattern}
         --outdir           : ${params.outdir}
         --ontreads         : ${params.ontreads}
         --title            : ${params.title}

         Runtime data:
        -------------------------------------------
         Running with profile:   ${ANSI_GREEN}${workflow.profile}${ANSI_RESET}
         Run container:          ${ANSI_GREEN}${workflow.container}${ANSI_RESET}
         Running as user:        ${ANSI_GREEN}${workflow.userName}${ANSI_RESET}
         Launch dir:             ${ANSI_GREEN}${workflow.launchDir}${ANSI_RESET}
         Base dir:               ${ANSI_GREEN}${baseDir}${ANSI_RESET}
         """
         .stripIndent()

def helpMessage() {
log.info """
        ===========================================
        SNP DETECTION F A S T Q C  TO VCF Table Generation P I P E L I N E
  
         Usage:
        -------------------------------------------
         --readsdir         : Directory with fastqc files, default is "fastqc"
         --fqpattern        : Regex pattern to match fastq files, default is "*R{1,2}*.fastq.gz"
         --outdir           : Where results will be saved, default is "results folder with sub folder of Modules process"
         --ontreads         : use this parameter for Nanopore reads
         --title            :Fatqc report title, default is "Summarized fastp report"
        ===========================================
         """
         .stripIndent()

}

//just in case trailing slash in readsdir not provided...
readsdir_repaired = "${params.readsdir}".replaceFirst(/$/, "/") 
//println(readsdir_repaired)

// build search pattern for fastq files in input dir
raw_data = readsdir_repaired + params.fqpattern
//println(raw_data)

// get counts of found fastq files
readcounts = file(raw_data)
println " Reads found:            ${ANSI_GREEN}${readcounts.size()}${ANSI_RESET}"
//println("$params.outdir")

process FASTQC {
        publishDir "$params.outdir/fastQC/beforeTrimming"
        input:
        tuple val(sample_id), path(reads)
        output:
        file("*.{html,zip}")
        script:
        """
        fastqc --noextract --nogroup --format 'fastq' --contaminants ${params.references}/Configuration/contaminant_list.txt ${reads[0]} ${reads[1]}
        """
}

process TRIMMOMATRIC {

    publishDir"$params.outdir/trimmomatic", mode: params.saveMode
    //container 'quay.io/biocontainers/trimmomatic:0.35--6'

    input:
    tuple val(genomeName),path(reads)

    output:
    tuple path(fq_1_paired), path(fq_2_paired),path(fq_1_unpaired), path(fq_2_unpaired) 

    script:

    fq_1_paired = genomeName + '_R1_trim_pair.fq'
    fq_1_unpaired = genomeName + '_R1_trim_unpair.fq'
    fq_2_paired = genomeName + '_R2_trim_pair.fq'
    fq_2_unpaired = genomeName + '_R2_trim_unpair.fq'

    """
    java -jar ${params.modules}/trimmomatic-0.39-2/trimmomatic.jar \
    PE -phred33 \
    ${reads[0]} \
    ${reads[1]} \
    $fq_1_paired \
    $fq_1_unpaired \
    $fq_2_paired \
    $fq_2_unpaired \
    ILLUMINACLIP:${params.modules}/trimmomatic-0.39-2/adapters/NexteraPE-PE.fa:2:30:10 SLIDINGWINDOW:4:20 LEADING:3 TRAILING:3 MINLEN:25
    """
}

process index_refseq {
        script:
        """
        bwa index ${params.references}/refseq/v1/all_seq.fa
        """
}

process FASTQC_AFTER_TRIM {
        publishDir "$params.outdir/fastQC/afterTrimming"
        input:
        tuple val(genomeName), path(reads)
        output:""
        script:
        """
        mkdir -p "$params.outdir/fastQC/afterTrimming"
        fastqc --noextract -t 55 --outdir $params.outdir/fastQC/afterTrimming  --nogroup --format 'fastq' --contaminants ${params.references}/Configuration/contaminant_list.txt $params.outdir/trimmomatic/${genomeName}_R1_trim_pair.fq $params.outdir/trimmomatic/${genomeName}_R1_trim_unpair.fq $params.outdir/trimmomatic/${genomeName}_R2_trim_unpair.fq $params.outdir/trimmomatic/${genomeName}_R2_trim_pair.fq
        """
        
}

process BWA_MEM {
        publishDir "$params.outdir/sam_files"
        input:
        tuple val(genomeName),path(reads)
        output: ""
        script:
        """
        mkdir -p "$params.outdir/sam_files"
        bwa mem -t 32 ${params.references}/refseq/v1/all_seq.fa $params.outdir/trimmomatic/${genomeName}_R1_trim_pair.fq $params.outdir/trimmomatic/${genomeName}_R2_trim_pair.fq > $params.outdir/sam_files/${genomeName}_R1_R2_001_trim_pair.bwa.sam 
        bwa mem -t 32 ${params.references}/refseq/v1/all_seq.fa $params.outdir/trimmomatic/${genomeName}_R1_trim_unpair.fq > $params.outdir/sam_files/${genomeName}_R1_001_trim_unpair.bwa.sam 
        bwa mem -t 32 ${params.references}/refseq/v1/all_seq.fa $params.outdir/trimmomatic/${genomeName}_R2_trim_unpair.fq > $params.outdir/sam_files/${genomeName}_R2_001_trim_unpair.bwa.sam
        """
}

reads_ch = Channel.fromFilePairs(raw_data, checkIfExists: true )

//println(reads_ch)
workflow {

    FASTQC(reads_ch)
    TRIMMOMATRIC(reads_ch) 
    index_refseq()
    
    FASTQC_AFTER_TRIM()
    
    
 

}

//=============================
workflow.onComplete {
    if (workflow.success) {
        

        log.info """
            ===========================================
            ${ANSI_GREEN}Finished in ${workflow.duration}
            Processed ${ readcounts.size() } fastq files
            See the fastQC Report here ==> ${ANSI_RESET}$params.outdir/fastQC/beforeTrimming
            """
            .stripIndent()
    } else {
        log.info """
            ===========================================
            ${ANSI_RED}Finished with errors!${ANSI_RESET}
            """
            .stripIndent()
    }
}

**Comment by mribeirodantas** [comment]
Hello, @satish15625. Next time, try to upload the script file instead of copy-pasting it in the text box. It's more practical for us to try to read/run your script 😄 

Yes, it is possible to force sequential runs. There is a nextflow pattern called [task batching](https://nextflow-io.github.io/patterns/task-batching/). In it, you handle numerous elements of a channel in a single task, which means they’ll be sequentially dealt with. That is, instead of creating 10 tasks to handle 1 element of your channel each, you can have one task taking care of the 10 elements at once, sequentially.

**Reply by satish15625** [reply]
Thanks @mribeirodantas  let me try this 
please check the attached file

**Reply by mribeirodantas** [reply]
Hello, Satish. There's no file attached here.

**Reply by satish15625** [reply]
[main.zip](https://github.com/nextflow-io/nextflow/files/10054771/main.zip)

**Reply by satish15625** [reply]
@mribeirodantas  can u check now

**Reply by mribeirodantas** [reply]
Hello, @satish15625. I see you're the same one asking this question in the Nextflow Slack, so I would rather keep the discussion there. Is this OK with you?

I gave you a solution over there, it's linked below. Can you please mark this as an answer so that this question appears as solved?

https://nextflow.slack.com/archives/C02T98A23U7/p1669042845376699?thread_ts=1669033491.659129&cid=C02T98A23U7

**Comment by jemunro** [comment]
You can also try using the maxForks directive, see: https://www.nextflow.io/docs/latest/process.html#maxforks

**Reply by mribeirodantas** [reply]
Oh, `maxForks`! It slipped my mind! Good one, @jemunro 😄
